### ISSUE TITLE
[BUG] Error: Failed to optimize the project: 'ShellCommand' object has no attribute '_program

### ISSUE THREAD

**Issue by cc-shekher** [duaraghav8/dockershrink#13]
**Describe the bug**
I am getting the same issue as described here https://github.com/duaraghav8/dockershrink/issues/1
I tried it for nodejs application only, I am using oven/bun image in my docker file with multiple targets. Isn't it supposed to be supported.

**To Reproduce**
```
dockershrink optimize
```

**Optional Screenshots**
<img width="649" alt="image" src="https://github.com/user-attachments/assets/637eecf6-252f-477f-9736-c099cc66d339">

**Comment by duaraghav8** [comment]
@cc-shekher there's no specific support for oven/bun (I haven't used it myself), but if ultimately it uses node.js, then I don't see a reason it shouldn't support. I'm going to check this in more detail.
Can you share your Dockerfile with me? Or at least a part of it over which I can reproduce this issue?

The next release will also contain a `--verbose` option which will print the whole stack trace that can be shared in bug reports.

**Comment by Creator54** [comment]
I am also facing the same issue, trying to create the smallest go runtime.

ref:
![image](https://github.com/user-attachments/assets/c2e357c9-5408-4292-be76-21f33e996c0d)

```Dockerfile
# First stage: Build a minimal Go toolchain
FROM golang@sha256:7313279ebd6d6509477292ff657a041dbf72a610f37804c70af82566a2c0493e AS builder

WORKDIR /go/src/app

# Install necessary tools
RUN apk add --no-cache binutils

# Create a minimal Go installation with only required components
RUN mkdir -p /go-minimal/go && \
    # Essential binaries and libraries
    mkdir -p /go-minimal/go/bin && \
    cp /usr/local/go/bin/go /go-minimal/go/bin/ && \
    # Essential pkg files (only what's needed for compilation)
    mkdir -p /go-minimal/go/pkg && \
    cp -r /usr/local/go/pkg/tool /go-minimal/go/pkg/ && \
    cp -r /usr/local/go/pkg/include /go-minimal/go/pkg/ && \
    # Essential standard library packages
    mkdir -p /go-minimal/go/src && \
    for pkg in fmt runtime builtin internal/race sync/atomic math math/bits \
        errors internal/fmtsort io internal/cpu unsafe os reflect sort strconv \
        unicode/utf8 sync internal/bytealg internal/goarch internal/goos \
        internal/reflectlite runtime/internal/atomic runtime/internal/math \
        runtime/internal/sys unicode internal/itoa syscall time path \
        internal/abi internal/unsafeheader internal/coverage/rtcov \
        internal/godebugs internal/goexperiment internal/poll \
        internal/safefilepath internal/syscall/execenv \
        internal/syscall/unix internal/testlog internal/oserror internal/bisect \
        internal/coverage internal/coverage/calloc internal/coverage/pods \
	internal/byteorder internal/chacha8rand internal/profilerecord internal/runtime/atomic \
	internal/runtime/exithook internal/runtime/maps internal/runtime/math internal/runtime/sys \
	internal/runtime/syscall internal/stringslite cmp iter internal/sync slices \
	internal/filepathlite internal/asan internal/msan internal/godebug \
        internal/coverage/stringtab internal/coverage/test internal/coverage/write; do \
        if [ -d "/usr/local/go/src/$pkg" ]; then \
            mkdir -p "/go-minimal/go/src/$(dirname $pkg)" && \
            cp -r "/usr/local/go/src/$pkg" "/go-minimal/go/src/$(dirname $pkg)/"; \
        fi \
    done && \
    # Copy platform-specific compiled packages
    GOOS=$(go env GOOS) && \
    GOARCH=$(go env GOARCH) && \
    PLATFORM_DIR="/usr/local/go/pkg/${GOOS}_${GOARCH}" && \
    if [ -d "$PLATFORM_DIR" ]; then \
        mkdir -p "/go-minimal/go/pkg/${GOOS}_${GOARCH}" && \
        for pkg in fmt.a runtime.a builtin.a math.a sync.a \
            internal/bytealg.a internal/goarch.a internal/goos.a \
            internal/reflectlite.a runtime/internal/atomic.a \
            runtime/internal/math.a runtime/internal/sys.a unicode.a \
            internal/itoa.a syscall.a time.a path.a internal/abi.a \
            internal/unsafeheader.a internal/coverage/rtcov.a \
            internal/godebugs.a internal/goexperiment.a internal/poll.a \
            internal/safefilepath.a internal/syscall/execenv.a \
            internal/syscall/unix.a internal/testlog.a internal/oserror.a; do \
            if [ -f "$PLATFORM_DIR/$pkg" ]; then \
                cp "$PLATFORM_DIR/$pkg" "/go-minimal/go/pkg/${GOOS}_${GOARCH}/"; \
            fi \
        done \
    fi && \
    # Strip debug information from binaries
    strip /go-minimal/go/bin/go

# Second stage: Create runtime image
FROM gcr.io/distroless/static-debian12:nonroot

# Copy only the minimal Go installation
COPY --from=builder /go-minimal/go /usr/local/go

# Set Go environment variables
ENV GOPATH=/go \
    PATH=$PATH:/usr/local/go/bin:/go/bin \
    GOROOT=/usr/local/go \
    HOME=/home/nonroot

# Create necessary directories with correct permissions
WORKDIR /go/src/app

USER nonroot

# Command to execute Go code passed during runtime
CMD ["/usr/local/go/bin/go", "run", "/go/src/app/main.go"]
```

**Comment by duaraghav8** [comment]
@Creator54 I'm Checking this

**Comment by duaraghav8** [comment]
I've figured out the root cause of this issue.
tldr; its a problem with Dockershrink's code and I'm working on a fix.

---

NOTES for me:

The issue occured specifically due to this code (you can write this in Dockerfile and run `dockershrink optimize --verbose` to reproduce)

```
for pkg in fmt runtime builtin internal/race sync/atomic math math/bits \
        errors internal/fmtsort io internal/cpu unsafe os reflect sort strconv \
        unicode/utf8 sync internal/bytealg internal/goarch internal/goos \
        internal/reflectlite runtime/internal/atomic runtime/internal/math \
        runtime/internal/sys unicode internal/itoa syscall time path \
        internal/abi internal/unsafeheader internal/coverage/rtcov \
        internal/godebugs internal/goexperiment internal/poll \
        internal/safefilepath internal/syscall/execenv \
        internal/syscall/unix internal/testlog internal/oserror internal/bisect \
        internal/coverage internal/coverage/calloc internal/coverage/pods \
	internal/byteorder internal/chacha8rand internal/profilerecord internal/runtime/atomic \
	internal/runtime/exithook internal/runtime/maps internal/runtime/math internal/runtime/sys \
	internal/runtime/syscall internal/stringslite cmp iter internal/sync slices \
	internal/filepathlite internal/asan internal/msan internal/godebug \
        internal/coverage/stringtab internal/coverage/test internal/coverage/write; do \
        if [ -d "/usr/local/go/src/$pkg" ]; then \
            mkdir -p "/go-minimal/go/src/$(dirname $pkg)" && \
            cp -r "/usr/local/go/src/$pkg" "/go-minimal/go/src/$(dirname $pkg)/"; \
        fi \
    done
```

This is because this code, when parsed into a node by `bashlex.parsesingle()`, results in an object with kind `CompoundNode`, which makes sense.
Currently, Dockershrink doesn't handle CompoundNode.
Because of this, the ShellCommand object's `_program` was never set. So when optimization logic tried to access cmd.program(), it threw exception.

The [shell_command parsing logic](https://github.com/duaraghav8/dockershrink/blob/main/dockershrink/dockerfile/shell_command.py#L169) needs to handle `compound` kind from the bash parser.

See bash documentation for [compound statements](https://www.gnu.org/software/bash/manual/html_node/Compound-Commands.html)

Code examples that are parsed as kind = `compound` by bashlex:

```python
import bashlex

code = "(echo hello)"
code = "(echo hello;)"
code = "(echo hello; echo world)"

code = """
if [ "$1" = "hello" ]; then
    echo "Greetings!"
else
    echo "No greeting provided."
fi
"""

tree = bashlex.parsesingle(code)
print(tree.kind)

code = """
for i in 1 2 3; do
    echo "for loop iteration: $i"
done
"""

tree = bashlex.parsesingle(code)
print(tree.kind)

code = """
while [ $COUNTER -lt 3 ]; do
    echo "while loop iteration: $COUNTER"
    ((COUNTER++))
done
"""

tree = bashlex.parsesingle(code)
print(tree.kind)

code = """
until [ $COUNTER2 -eq 0 ]; do
    echo "until loop iteration: $COUNTER2"
    ((COUNTER2--))
done
"""

tree = bashlex.parsesingle(code)
print(tree.kind)
```

**TODO**: How to handle compound node? It could be a loop statement, if conditional or something else. Calling ShellCommand.program() on a CompoundNode doesn't make sense, there is no program at the start but more commands inside that statement.

### Strategy
Compound statements, as seen above, are loop constructs, conditionals & chained commands.
All these are more complex than just single bash commands and it is difficult for the rule engine to determine the intentions of these logical statements.
So we want to let the LLM deal with compound statements.
For parsing and rule engine, we should simply "recognize and ignore" compound nodes so that they don't break the app.

**Comment by duaraghav8** [comment]
closing this issue as dockershrink is now ported to Golang and this won't be occuring anymore.
